cmd/query/app/querysvc/query_service_v2.go

```diff
@@ -0,0 +1,140 @@
+type QueryServiceOptionsV2 struct {
+	ArchiveTraceReader tracestore.Reader
+	ArchiveTraceWriter tracestore.Writer
+	Adjuster           adjuster.Adjuster
+}
+
+// StorageCapabilities is a feature flag for query service
+type StorageCapabilitiesV2 struct {
+	ArchiveStorage bool `json:"archiveStorage"`
+	// TODO: Maybe add metrics Storage here
+	// SupportRegex     bool
+	// SupportTagFilter bool
+}
+
+// QueryService contains span utils required by the query-service.
+type QueryServiceV2 struct {
+	traceReader      tracestore.Reader
+	dependencyReader depstore.Reader
+	options          QueryServiceOptionsV2
+}
+
+// NewQueryService returns a new QueryService.
+func NewQueryServiceV2(
+	traceReader tracestore.Reader,
+	dependencyReader depstore.Reader,
+	options QueryServiceOptionsV2,
+) *QueryServiceV2 {
+	qsvc := &QueryServiceV2{
+		traceReader:      traceReader,
+		dependencyReader: dependencyReader,
+		options:          options,
+	}
+
+	if qsvc.options.Adjuster == nil {
+		qsvc.options.Adjuster = adjuster.Sequence(adjuster.StandardAdjusters(defaultMaxClockSkewAdjust)...)
+	}
+	return qsvc
+}
+
+// GetTrace is the queryService implementation of tracestore.Reader.GetTrace
+func (qs QueryServiceV2) GetTraces(ctx context.Context, traceIDs ...tracestore.GetTraceParams) iter.Seq2[[]ptrace.Traces, error] {
+	traceIter := qs.traceReader.GetTraces(ctx, traceIDs...)
+	_, err := iter.FlattenWithErrors(traceIter)
+	if errors.Is(err, spanstore.ErrTraceNotFound) {
+		if qs.options.ArchiveTraceReader == nil {
+			return func(yield func([]ptrace.Traces, error) bool) {
+				yield(nil, err)
+			}
+		}
+		traceIter = qs.options.ArchiveTraceReader.GetTraces(ctx, traceIDs...)
+	}
+	return traceIter
+}
+
+// GetServices is the queryService implementation of tracestore.Reader.GetServices
+func (qs QueryServiceV2) GetServices(ctx context.Context) ([]string, error) {
+	return qs.traceReader.GetServices(ctx)
+}
+
+// GetOperations is the queryService implementation of tracestore.Reader.GetOperations
+func (qs QueryServiceV2) GetOperations(
+	ctx context.Context,
+	query tracestore.OperationQueryParameters,
+) ([]tracestore.Operation, error) {
+	return qs.traceReader.GetOperations(ctx, query)
+}
+
+// FindTraces is the queryService implementation of tracestore.Reader.FindTraces
+func (qs QueryServiceV2) FindTraces(ctx context.Context, query tracestore.TraceQueryParams) iter.Seq2[[]ptrace.Traces, error] {
+	return qs.traceReader.FindTraces(ctx, query)
+}
+
+// ArchiveTrace is the queryService utility to archive traces.
+func (qs QueryServiceV2) ArchiveTrace(ctx context.Context, traceID pcommon.TraceID) error {
+	if qs.options.ArchiveTraceWriter == nil {
+		return errNoArchiveSpanStorage
+	}
+	var err error
+	traces, err := iter.FlattenWithErrors(qs.GetTraces(ctx, tracestore.GetTraceParams{TraceID: traceID}))
+	if err != nil {
+		return err
+	}
+	for _, trace := range traces {
+		err = errors.Join(err, qs.options.ArchiveTraceWriter.WriteTraces(ctx, trace))
+	}
```

Comment:
@yurishkuro do we need to loop here or should we just be taking the first element?

Reply:
We need to loop through `Seq` and call `write` on each chunk. Flatten was never meant to be used in prod v2 code, it was just a helper useful in v1 bridge, since v1 is not streaming.